src/pages/DataAccessRequestApplication.js

```diff
@@ -481,43 +556,87 @@ class DataAccessRequestApplication extends Component {
     this.setState({ showDialogSave: true });
   };
 
-  dialogHandlerSubmit = (answer) => (e) => {
+  //Can't do uploads in parallel since endpoints are post and they both alter attributes in json column
+  //If done in parallel, updated attribute of one document will be overwritten by the outdated value on the other
+  saveDULDocuments = async(uploadedIrbDocument = null, uploadedCollaborationLetter = null, referenceId) => {
+    let irbUpdate, collaborationUpdate;
+    irbUpdate = await DAR.uploadDULDocument(uploadedIrbDocument, referenceId, 'irbDocument');
+    collaborationUpdate = await DAR.uploadDULDocument(uploadedCollaborationLetter, referenceId, 'collaborationDocument');
+    return Object.assign({}, irbUpdate.data, collaborationUpdate.data);
+  }
+
+  updateDraftResponse = (formattedFormData, referenceId) => {
+    let darPartialResponse;
+    if(!isNil(referenceId) && !isEmpty(referenceId)) {
+      darPartialResponse = DAR.updateDarDraft(formattedFormData, referenceId);
+    } else {
+      darPartialResponse = DAR.postDarDraft(formattedFormData);
+    }
+    return darPartialResponse;
+  }
+
+  submitDARFormData = async (answer) => {
     if (answer === true) {
-      let ontologies = [];
+      const userId = Storage.getCurrentUser().dacUserId;
+      const {uploadedIrbDocument, uploadedCollaborationLetter} = this.state.step2;
+      let formattedFormData = fp.cloneDeep(this.state.formData);
+      const ontologies = fp.map((item) => {
+        return {
+          id: item.key,
+          value: item.value
+        };
+      })(formattedFormData.ontologies);
+
       for (let ontology of this.state.formData.ontologies) {
         ontologies.push(ontology.item);
       }
-      this.setState(prev => {
-        if (ontologies.length > 0) {
-          prev.formData.ontologies = ontologies;
-        }
-        for (var key in prev.formData) {
-          if (prev.formData[key] === '') {
-            prev.formData[key] = undefined;
-          }
+      if (ontologies.length > 0) {
+        formattedFormData.ontologies = ontologies;
+      }
+      for (var key in formattedFormData) {
+        if (formattedFormData[key] === '') {
+          formattedFormData[key] = undefined;
         }
-        return prev;
-      }, () => {
-        let formData = this.state.formData;
-        // DAR datasetIds needs to be a list of ids
-        formData.datasetIds = fp.map('value')(formData.datasets);
-        formData.userId = Storage.getCurrentUser().dacUserId;
-        this.setState(prev => {
-          prev.disableOkBtn = true;
-          return prev;
+      }
+      formattedFormData.datasetIds = fp.map('value')(formattedFormData.datasets);
+      formattedFormData.userId = userId;
+
+      try {
+        //NOTE: the pre-processing saves are adding time to record generation (makes front-end seem slow)
```

Comment:
We should look at improving the endpoint performance if this is affecting the UI. It really shouldn't, we don't do a lot of work when saving a draft DAR. Will file a story for that.

Reply:
Created https://broadinstitute.atlassian.net/browse/DUOS-835 to look at this.